### PR TITLE
Refactoring Ui events in core

### PR DESCRIPTION
**Pull request by moniika**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves
 Part of #4203

### Proposed Changes

- Moves event code into directory `core/events/`
- Adds new property to `Blockly.Events.Abstract` of `isUiEvent` in order to support check `event.isUiEvent` instead of using `event.type == 'ui'`
- Adds a new class `Blockly.Events.UiBase`
- Refactors all Ui events in core to extend from it in a custom class (changing the `type` for each event and changing properties associated with each event) but left old `Blockly.Events.Ui` class with the same properties.
- Updates logic in core to reflect changes in Ui event properties

- Adds tests for serialization of events and includes some fixes for event serialization

| class | type | properties | old ui events |
| --- | --- | --- | --- |
| Blockly.Events.Drag | drag | blocks, isStart | dragStart, dragStop |
| Blockly.Events.Selected | selected | oldElementId, elementId | select | 
| Blockly.Events.Click | click | targetType | click, zoom | 
| Blockly.Events.MarkerMove | marker_move | oldNode, curNode, isCursor | cursorMove, markerMove | 
| Blockly.Events.BubbleOpen | bubble_open | isOpen, element | commentOpen, mutatorOpen, warningOpen | 
| Blockly.Events.TrashcanOpen | trashcanOpen | isOpen | trashcanOpen | 
| Blockly.Events.ToolboxItemSelect | toolbox_item_change | oldItem, newItem | category | 
| Blockly.Events.ThemeChange | theme_change | themeName | theme | 
| Blockly.Events.ViewportChange | viewport_change | top, left, scale | (new event) | 

### Reason for Changes

Explained in issue.

### Test Coverage

Ran mocha tests and checked using playground that expected events were emitted.

Tested on:
 * Desktop Chrome

### Documentation

Documentation for events will need to be updated:
https://developers.google.com/blockly/guides/configure/web/events

And clear instructions for what changed need to be done when updating to new release.

### Additional Information

This is a breaking change for listening to UI events emitted by Blockly core, but should not break anyone that extended from the old UI event class.

### CODE REVIEW

File: core/events/ui_events.js [hunk 1]
```diff
@@ -11,42 +11,73 @@
...
+  // UI events do not undo or redo.
+  this.recordUndo = false;
+};
+Blockly.utils.object.inherits(Blockly.Events.UiBase, Blockly.Events.Abstract);
+
+/**
+ * Whether or not the event is a UI event.
+ * @type {boolean}
+ */
+Blockly.Events.UiBase.prototype.IS_UI_EVENT = true;
```
Comment (alschmiedt) [inline core/events/ui_events.js:56]: Could this get set in the constructor?

File: core/events/ui_events.js [hunk 2]
```diff
@@ -11,42 +11,73 @@
...
+ * Class for a UI event.
  * @param {?Blockly.Block=} opt_block The affected block.  Null for UI events
  *     that do not have an associated block.  Undefined for a blank event.
  * @param {string=} opt_element One of 'selected', 'comment', 'mutatorOpen',
  *     etc.
  * @param {*=} opt_oldValue Previous value of element.
  * @param {*=} opt_newValue New value of element.
- * @extends {Blockly.Events.Abstract}
+ * @extends {Blockly.Events.UiBase}
+ * @deprecated Instead use a more specific UI event.
```
Comment (alschmiedt) [inline core/events/ui_events.js:67]: Add when it was deprecated so we know approximately when we should remove.

**Review by alschmiedt — APPROVED** [review]
A few nits, but overall lgtm

File: core/events/ui_events.js (hunk 2 above)
Comment (moniika) [reply core/events/ui_events.js]: Done.

File: core/events/ui_events.js (hunk 1 above)
Comment (moniika) [reply core/events/ui_events.js]: I'd prefer to set it here on the prototype since in shouldn't change per instance of class.

**Pull request merged**